### ISSUE TITLE
dstNode is undefined in function prepareNodeCopyAsDragImage

### ISSUE THREAD

**Issue by z2oh** [timruffles/ios-html5-drag-drop-shim#111]
In the function [`prepareNodeCopyAsDragImage`](https://github.com/timruffles/ios-html5-drag-drop-shim/blob/b485129b36fb875a072743df1a1d641b4183c34d/src/index.ts#L1270) I am running into an error that causes dragging and dropping to break. In an attempt to debug this issue I found that `dstNode` was sometimes undefined, which would break the script on 

```dstNode.style.setProperty(...);```

Sometimes, `dstNode` was a text node (`dstNode.nodeType === 3`) which would also break on the above line, as the `style` object was undefined. I was able to 'fix' this by wrapping all the problem lines in

```
if(dstNode && dstNode.style) {
    ...
}
```

but I'm not sure if this just fixes a symptom of some other root cause.

I am using Polymer to polyfill webcomponents, which has made it difficult to locate exactly what HTML was causing this to break (and difficult to post steps to reproduce). I was able to narrow it down to the [`<paper-icon-button>`](https://github.com/polymerelements/paper-icon-button) element, which is mostly just a wrapper around an SVG. When I isolate the SVG, the dragging works fine without my hotfix  so I am stumped as to what undefined node is causing this to break. Any ideas on how I can debug this further? I am happy to open a PR once I have a better understanding of the underlying issue.

**Comment by reppners** [comment]
Thanks for the detailed report and your effort in investigation. I don't have time right now to go deep but recommend investigating further on the result of https://github.com/timruffles/ios-html5-drag-drop-shim/blob/b485129b36fb875a072743df1a1d641b4183c34d/src/index.ts#L1303

Because in your case obviously the `cloneNode()` method does not return a result that is a matching clone I'd loop over srcNode and dstNode in one loop just printing the tree and highlighting the differences. This might indicate on which parts of the srcNode tree cloning did fail.

Maybe there is some bug or edge cases regarding the cloneNode() method that is rather undocumented and the polyfill trips over it.

**Comment by reppners** [comment]
Thanks for posting your findings.

I'd rather push for a solution that works with webcomponents/shadowdom.

So based on more research I think we can find a solution that can be included into the library. Don't have any knowledge about cloning and shadow dom - maybe there's some documentation.

Another approach could be to attach the node to the DOM with `visibility: hidden` or `display:none` so it initializes and only after that transfer the styles.

**Comment by reppners** [comment]
I think the problem originates from ShadyDOM and cloning without using the ShadyDOM API. Here is a related issue https://github.com/Polymer/polymer/issues/1888

**Comment by z2oh** [comment]
It does seem related to that issue, although when replacing https://github.com/timruffles/ios-html5-drag-drop-shim/blob/0f2c6426d2618fadd1e9dfa4816fa0384b0291a0/release/index.js#L604 with

```
var dragImage = Polymer.dom(sourceNode).cloneNode(true);
```

I run into the same issue. When running through the chrome debugger I notice that all the ShadyDOM nodes are there. However, I stepped through the code and found this (notice the evaluated `srcNode` and `dstNode`):

![textnode](https://user-images.githubusercontent.com/385684/30082076-866568f6-9257-11e7-99a7-d5cc308c824a.png)

It would seem the svg node is incorrectly cloned as a text node? Text nodes do not have a style property so an error occurs on line 590 in the screenshot above.

I will do some research to see if this issue still occurs when there are no SVGs in the ShadyDOM.

**Comment by z2oh** [comment]
I tried nesting some components and making the parent draggable, and the drag image shows up just fine including all the child elements that are in ShadyDOM. It looks like this issue may be somehow isolated to SVGs in ShadyDOM specifically.

**Comment by reppners** [comment]
API additions to enable custom dragImageSetup released as of v2.3.0-rc.0. Closing this one as the polyfill can't yet implement a general purpose solution for ShadowDOM/ShadyDOM.
